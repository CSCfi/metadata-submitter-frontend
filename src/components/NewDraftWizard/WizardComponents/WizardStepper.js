//@flow
import React, { useState } from "react"

import Button from "@material-ui/core/Button"
import Step from "@material-ui/core/Step"
import StepConnector from "@material-ui/core/StepConnector"
import StepLabel from "@material-ui/core/StepLabel"
import Stepper from "@material-ui/core/Stepper"
import { makeStyles, withStyles } from "@material-ui/core/styles"
import ArrowBackIosIcon from "@material-ui/icons/ArrowBackIos"
import ArrowForwardIosIcon from "@material-ui/icons/ArrowForwardIos"
import Check from "@material-ui/icons/Check"
import clsx from "clsx"
import { useDispatch, useSelector } from "react-redux"

import WizardAlert from "./WizardAlert"

import { resetDraftStatus } from "features/draftStatusSlice"
import { resetObjectType } from "features/wizardObjectTypeSlice"
import { decrement, increment } from "features/wizardStepSlice"
import { resetSubmissionType } from "features/wizardSubmissionTypeSlice"
import type { CreateFolderFormRef } from "types"
/*
 * Customized stepper inspired by https://material-ui.com/components/steppers/#customized-stepper
 */
const QontoConnector = withStyles(theme => ({
  alternativeLabel: {
    top: 10,
    left: "calc(-50% + 16px)",
    right: "calc(50% + 16px)",
  },
  active: {
    "& $line": {
      borderColor: theme.palette.primary.main,
    },
  },
  completed: {
    "& $line": {
      borderColor: theme.palette.primary.main,
    },
  },
  line: {
    borderColor: theme.palette.secondary.main,
    borderTopWidth: 3,
    borderRadius: 1,
  },
}))(StepConnector)

const useQontoStepIconStyles = makeStyles(theme => ({
  root: {
    color: theme.palette.secondary.main,
    display: "flex",
    height: 22,
    alignItems: "center",
  },
  active: {
    color: theme.palette.primary.main,
  },
  circle: {
    width: 8,
    height: 8,
    borderRadius: "50%",
    backgroundColor: "currentColor",
  },
  completed: {
    color: theme.palette.primary.main,
    zIndex: 1,
    fontSize: 18,
  },
  floating: {
    border: "solid 1px #000",
    backgroundColor: "#FFF",
    boxShadow: 0,
  },
}))

function QontoStepIcon(props: { active: boolean, completed: boolean }) {
  const classes = useQontoStepIconStyles()
  const { active, completed } = props

  return (
    <div
      className={clsx(classes.root, {
        [classes.active]: active,
      })}
    >
      {completed ? <Check className={classes.completed} /> : <div className={classes.circle} />}
    </div>
  )
}

const useStyles = makeStyles({
  stepper: {
    display: "flex",
    flexDirection: "row",
    width: "100%",
    borderBottom: "solid 1px #ccc",
  },
  stepperContainer: {
    width: "80%",
  },
  centeredStepButton: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 0,
    border: "none",
    width: "10%",
    backgroundColor: "rgba(139, 26, 79, 0.04)",
    "&:hover": {
      border: "none",
      backgroundColor: "rgba(139, 26, 79, 0.16)",
    },
    "&.Mui-disabled": {
      border: "none",
    },
  },
})

/**
 * Show info about wizard steps to user.
 * If createFolderForm is passed as reference it is used to trigger correct form when clicking next.
 */
const WizardStepper = ({ createFolderFormRef }: { createFolderFormRef?: CreateFolderFormRef }): React$Element<any> => {
  const classes = useStyles()
  const dispatch = useDispatch()
  const wizardStep = useSelector(state => state.wizardStep)
  const steps = ["Folder Name & Description", "Add Objects", "Summary"]
  const formState = useSelector(state => state.submissionType)
  const [alert, setAlert] = useState(false)
  const [direction, setDirection] = useState("")
  const draftStatus = useSelector(state => state.draftStatus)

  const handleNavigation = (step: boolean) => {
    setDirection("")
    setAlert(false)
    dispatch(resetDraftStatus())
    if (step) {
      direction === "previous" ? dispatch(decrement()) : dispatch(increment())
      dispatch(resetObjectType())
      dispatch(resetSubmissionType())
    }
  }

  return (
    <div className={classes.stepper}>
      <Button
        className={classes.centeredStepButton}
        disableElevation
        color="primary"
        variant="outlined"
        disabled={wizardStep < 1}
        onClick={() => {
          if (wizardStep === 1 && formState.trim().length > 0 && draftStatus === "notSaved") {
            setDirection("previous")
            setAlert(true)
          } else {
            dispatch(decrement())
          }
        }}
      >
        <ArrowBackIosIcon fontSize="large" />
        Back
      </Button>
      <Stepper
        activeStep={wizardStep}
        className={classes.stepperContainer}
        alternativeLabel
        connector={<QontoConnector />}
      >
        {steps.map(label => (
          <Step key={label}>
            <StepLabel StepIconComponent={QontoStepIcon}>{label}</StepLabel>
          </Step>
        ))}
      </Stepper>
      <Button
        disabled={createFolderFormRef?.current?.isSubmitting || wizardStep >= 2}
        className={classes.centeredStepButton}
        disableElevation
        color="primary"
        variant="outlined"
        onClick={async () => {
          if (createFolderFormRef?.current) {
            await createFolderFormRef.current.dispatchEvent(new Event("submit", { cancelable: true }))
          }
          if (wizardStep === 1 && formState.trim().length > 0 && draftStatus === "notSaved") {
            setDirection("next")
            setAlert(true)
          } else if (wizardStep !== 2 && !createFolderFormRef?.current) {
            dispatch(increment())
          }
        }}
      >
        Next
        <ArrowForwardIosIcon fontSize="large" />
      </Button>
      {wizardStep === 1 && alert && (
        <WizardAlert onAlert={handleNavigation} parentLocation="stepper" alertType={direction}></WizardAlert>
      )}
    </div>
  )
}

export default WizardStepper
