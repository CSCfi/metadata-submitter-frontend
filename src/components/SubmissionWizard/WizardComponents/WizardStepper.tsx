import React, { useState } from "react"

import ArrowBackIosIcon from "@mui/icons-material/ArrowBackIos"
import ArrowForwardIosIcon from "@mui/icons-material/ArrowForwardIos"
import Check from "@mui/icons-material/Check"
import Button from "@mui/material/Button"
import Step from "@mui/material/Step"
import StepConnector from "@mui/material/StepConnector"
import StepLabel from "@mui/material/StepLabel"
import Stepper from "@mui/material/Stepper"
import { makeStyles, withStyles } from "@mui/styles"
import clsx from "clsx"
import { useNavigate, useParams } from "react-router-dom"

import WizardAlert from "./WizardAlert"

import { resetDraftStatus } from "features/draftStatusSlice"
import { resetObjectType } from "features/wizardObjectTypeSlice"
import { resetSubmissionType } from "features/wizardSubmissionTypeSlice"
import { useAppSelector, useAppDispatch } from "hooks"
import { useQuery, pathWithLocale } from "utils"
/*
 * Customized stepper inspired by https://material-ui.com/components/steppers/#customized-stepper
 */
const QontoConnector = withStyles(theme => ({
  alternativeLabel: {
    top: 10,
    left: `calc(-50% + ${8})`,
    right: `calc(50% + ${8})`,
  },
  active: {
    "& $line": {
      borderColor: theme.palette.primary.main,
    },
  },
  completed: {
    "& $line": {
      borderColor: theme.palette.primary.main,
    },
  },
  line: {
    borderColor: theme.palette.secondary.main,
    borderTopWidth: 3,
    borderRadius: 1,
  },
}))(StepConnector)

const useQontoStepIconStyles = makeStyles(theme => ({
  root: {
    color: theme.palette.secondary.main,
    display: "flex",
    height: 22,
    alignItems: "center",
  },
  active: {
    color: theme.palette.primary.main,
  },
  circle: {
    width: 8,
    height: 8,
    borderRadius: "50%",
    backgroundColor: "currentColor",
  },
  completed: {
    color: theme.palette.primary.main,
    zIndex: 1,
    fontSize: 18,
  },
  floating: {
    border: "solid 1px #000",
    backgroundColor: theme.palette.background.default,
    boxShadow: "0",
  },
}))

function QontoStepIcon(props: { active: boolean; completed: boolean }) {
  const classes = useQontoStepIconStyles()
  const { active, completed } = props

  return (
    <div
      className={clsx(classes.root, {
        [classes.active]: active,
      })}
    >
      {completed ? <Check className={classes.completed} /> : <div className={classes.circle} />}
    </div>
  )
}

const useStyles = makeStyles({
  stepper: {
    display: "flex",
    flexDirection: "row",
    width: "100%",
    borderBottom: "solid 1px #ccc",
    marginTop: 50,
  },
  stepperContainer: {
    width: "80%",
  },
  centeredStepButton: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 0,
    border: "none",
    width: "10%",
    backgroundColor: "rgba(139, 26, 79, 0.04)",
    "&:hover": {
      border: "none",
      backgroundColor: "rgba(139, 26, 79, 0.16)",
    },
    "&.Mui-disabled": {
      border: "none",
    },
  },
})

/**
 * Show info about wizard steps to user.
 * If createFolderForm is passed as reference it is used to trigger correct form when clicking next.
 */

const WizardStepper = ({
  createFolderFormRef,
}: {
  createFolderFormRef?: { current: HTMLElement | null } | undefined
}) => {
  const classes = useStyles()
  const dispatch = useAppDispatch()
  const steps = ["Folder Name & Description", "Add Objects", "Summary"]
  const formState = useAppSelector(state => state.submissionType)
  const [alert, setAlert] = useState(false)
  const [direction, setDirection] = useState("")
  const draftStatus = useAppSelector(state => state.draftStatus)
  const navigate = useNavigate()
  const params = useParams()

  const folderId = params.folderId
  const newDraftPath = pathWithLocale(`submission/${folderId}`)

  const queryParams = useQuery()
  const wizardStep = Number(queryParams.get("step"))

  const unsavedSubmission = wizardStep === 1 && formState.trim().length > 0 && draftStatus === "notSaved"

  const handleNavigation = (step: boolean) => {
    setDirection("")
    setAlert(false)
    dispatch(resetDraftStatus())

    if (step) {
      direction === "previous" ? navigate(-1) : navigate({ pathname: newDraftPath, search: "step=2" })
      dispatch(resetObjectType())
      dispatch(resetSubmissionType())
    }
  }

  return (
    <div className={classes.stepper}>
      <Button
        className={classes.centeredStepButton}
        disableElevation
        color="primary"
        variant="outlined"
        disabled={wizardStep < 1}
        onClick={() => {
          if (unsavedSubmission) {
            setDirection("previous")
            setAlert(true)
          } else {
            navigate({ pathname: newDraftPath, search: `step=${wizardStep - 1}` })
          }
        }}
      >
        <ArrowBackIosIcon fontSize="large" />
        Back
      </Button>
      <Stepper
        activeStep={wizardStep}
        className={classes.stepperContainer}
        alternativeLabel
        connector={<QontoConnector />}
      >
        {steps.map(label => (
          <Step key={label}>
            <StepLabel StepIconComponent={QontoStepIcon}>{label}</StepLabel>
          </Step>
        ))}
      </Stepper>
      <Button
        disabled={wizardStep >= 2}
        className={classes.centeredStepButton}
        disableElevation
        color="primary"
        variant="outlined"
        onClick={async () => {
          if (createFolderFormRef?.current) {
            await createFolderFormRef.current.dispatchEvent(new Event("submit", { cancelable: true, bubbles: true }))
          }
          if (unsavedSubmission) {
            setDirection("next")
            setAlert(true)
          } else if (wizardStep !== 2 && !createFolderFormRef?.current) {
            navigate({ pathname: newDraftPath, search: "step=2" })
          }
        }}
      >
        Next
        <ArrowForwardIosIcon fontSize="large" />
      </Button>
      {wizardStep === 1 && alert && (
        <WizardAlert onAlert={handleNavigation} parentLocation="stepper" alertType={direction}></WizardAlert>
      )}
    </div>
  )
}

export default WizardStepper
