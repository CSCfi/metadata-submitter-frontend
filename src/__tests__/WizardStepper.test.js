import React from "react"

import "@testing-library/jest-dom/extend-expect"
import { render, cleanup, screen } from "@testing-library/react"
import { Provider } from "react-redux"
import configureStore from "redux-mock-store"

import WizardStepper from "../components/NewDraftWizard/WizardComponents/WizardStepper"

const mockStore = configureStore([])

afterEach(cleanup)

describe("WizardStepper", () => {
  const store = mockStore({
    submissionType: "",
    wizardStep: 0,
  })
  it("should take a snapshot", () => {
    const wrapper = (
      <Provider store={store}>
        <WizardStepper />
      </Provider>
    )
    const { asFragment } = render(wrapper)
    expect(asFragment(wrapper)).toMatchSnapshot()
  })

  it("should have back-step button disabled when wizardStep is 0", () => {
    const wrapper = (
      <Provider store={store}>
        <WizardStepper />
      </Provider>
    )
    const { getByText } = render(wrapper)
    expect(getByText("Back").closest("Button")).toHaveClass("Mui-disabled")
  })

  it("should add to wizardStep store item on next-button click", () => {
    const wrapper = (
      <Provider store={store}>
        <WizardStepper createFolderFormRef={[]} />
      </Provider>
    )
    const { getByText } = render(wrapper)
    // fireEvent.click(getByText("Next").closest("Button"))
    const button = getByText("Next").closest("Button")
    screen.debug(button)
    // console.log(button)
    // fireEvent.click(button)
    // // expect(button).toHaveBeenCalled()
    // expect(store.wizardStep).toBe(1)
  })
})
