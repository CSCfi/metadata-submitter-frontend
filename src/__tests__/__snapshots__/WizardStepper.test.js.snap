// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardStepper should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-stepper-1"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-centeredStepButton-3 MuiButton-outlinedPrimary MuiButton-disableElevation Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
        Back
      </span>
    </button>
    <div
      class="MuiPaper-root MuiStepper-root MuiStepper-horizontal makeStyles-stepperContainer-2 MuiStepper-alternativeLabel MuiPaper-elevation0"
    >
      <div
        class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel"
      >
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal MuiStepLabel-alternativeLabel"
        >
          <span
            class="MuiStepLabel-iconContainer MuiStepLabel-alternativeLabel"
          >
            <div
              class="makeStyles-root-4 makeStyles-active-5"
            >
              <div
                class="makeStyles-circle-6"
              />
            </div>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiStepLabel-alternativeLabel MuiStepLabel-active MuiTypography-body2 MuiTypography-displayBlock"
            >
              Folder Name & Description
            </span>
          </span>
        </span>
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel"
      >
        <div
          class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel WithStyles(ForwardRef(StepConnector))-alternativeLabel-9 Mui-disabled"
        >
          <span
            class="MuiStepConnector-line WithStyles(ForwardRef(StepConnector))-line-12 MuiStepConnector-lineHorizontal"
          />
        </div>
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel"
        >
          <span
            class="MuiStepLabel-iconContainer MuiStepLabel-alternativeLabel"
          >
            <div
              class="makeStyles-root-4"
            >
              <div
                class="makeStyles-circle-6"
              />
            </div>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiStepLabel-alternativeLabel MuiTypography-body2 MuiTypography-displayBlock"
            >
              Add Objects
            </span>
          </span>
        </span>
      </div>
      <div
        class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel"
      >
        <div
          class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel WithStyles(ForwardRef(StepConnector))-alternativeLabel-9 Mui-disabled"
        >
          <span
            class="MuiStepConnector-line WithStyles(ForwardRef(StepConnector))-line-12 MuiStepConnector-lineHorizontal"
          />
        </div>
        <span
          class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel"
        >
          <span
            class="MuiStepLabel-iconContainer MuiStepLabel-alternativeLabel"
          >
            <div
              class="makeStyles-root-4"
            >
              <div
                class="makeStyles-circle-6"
              />
            </div>
          </span>
          <span
            class="MuiStepLabel-labelContainer"
          >
            <span
              class="MuiTypography-root MuiStepLabel-label MuiStepLabel-alternativeLabel MuiTypography-body2 MuiTypography-displayBlock"
            >
              Summary
            </span>
          </span>
        </span>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-centeredStepButton-3 MuiButton-outlinedPrimary MuiButton-disableElevation"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Next
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
